lib/CL/devices/level0/level0-driver.cc

```diff
@@ -684,28 +701,38 @@ void *Level0Queue::createCommandBuffer(cl_command_buffer_khr CmdBuf) {
     Cmd->device = nullptr;
   }
 
-  makeMemResident();
   syncMemHostPtrs();
   std::queue<ze_event_handle_t> CmdBufEvtList;
   closeCmdList(&CmdBufEvtList);
 
   POCL_MEASURE_FINISH(ZeListPrepare);
   std::swap(CmdListH, SaveCmdListH);
   Level0CmdBufferData *CmdBufData = new Level0CmdBufferData;
+  assert(CmdBufData);
   CmdBufData->CmdListH = SaveCmdListH;
   CmdBufData->Events.swap(CmdBufEvtList);
+  CmdBufData->MemPtrsToMakeResident.swap(MemPtrsToMakeResident);
+  CurrentEventH = nullptr;
+  PreviousEventH = nullptr;
+  assert(CmdListH != nullptr);
+  assert(UseMemHostPtrsToSync.empty());
+  assert(MemPtrsToMakeResident.empty());
+  assert(DeviceEventsToReset.empty());
   return (void *)CmdBufData;
 }
 
 void Level0Queue::freeCommandBuffer(void *CmdBufPtr) {
   assert(CmdBufPtr);
   Level0CmdBufferData *CmdBufData = (Level0CmdBufferData *)CmdBufPtr;
-  while (!CmdBufData->Events.empty()) {
-    auto E = CmdBufData->Events.front();
-    CmdBufData->Events.pop();
-    AvailableDeviceEvents.push(E);
+  {
```

Comment:
Do we need the separate code block here for the lock guard? Could it be function scope?

Reply:
No, it can't be function scope. The delete operator deletes the CmdBuf object ; we can't be holding a lock that's inside the object, because the lock would be unlocked at function exit (after delete), creating a write-after-free.